### DISCUSSION TITLE
How to break the multiple steps macro from userscript?

### DISCUSSION THREAD

**Discussion by laughtingman** [chhoumann/quickadd#753] [Q&A] [answered]
I have a macro with multiple steps:
1. run script
2. create note from template
3. rum obsidian command
4. something else

At the first step I have userscript with something like this:
```javascript
async function start(params, settings) {
   QuickAdd = params;
   Settings = settings;
   const query = await QuickAdd.quickAddApi.inputPrompt("Insert a query: ");
	//if I click 'cancel', 'esc', or stay input empty
        if (!query) { 
            notice("You don't enter query");
            throw new Error("Empty query");
            //I whant to break and prevent all next macro steps in this case. How?
        }
       //next logic...
}
```

How to stop the macro from the code and prevent to do steps 2-4?

**Comment by chhoumann** [comment, marked as answer]
Coming back to this one late - sorry about that.

There's a first-class way to do exactly this now: `params.abort()`, shipped in QuickAdd 2.10.0.

```javascript
async function start(params, settings) {
    const query = await params.quickAddApi.inputPrompt("Insert a query: ");
    // cancel, esc, or empty input
    if (!query) {
        params.abort("You didn't enter a query");
    }
    // next logic...
}
```

`params.abort(message)` stops the macro immediately - steps 2-4 never run - shows your message as a notice, and doesn't pop an error dialog. (An uncaught `throw` like in your snippet also halts the remaining steps these days, but it gets reported as a script error; `abort` is the clean cancel.)

Docs: https://quickadd.obsidian.guide/docs/Choices/MacroChoice (see "Macro Execution Control").

Closing as resolved.
